### PR TITLE
[Spot] Allow 2x spot jobs on a spot controller

### LINKED ISSUE #3187: [Spot] Allow more spot jobs on a controller
It has been observed that changing the `DEFAULT_TASK_CPU_DEMAND` to `0.2` will work for submitting a significant amount of spot jobs, we should change the CPU requirement for each spot controller process to `0.2` or `0.25` (the same as serve controller process) to allow more spot jobs on a controller.

https://github.com/skypilot-org/skypilot/blob/1c063fb71083d4ac4a3825a9af7295815d077d51/sky/backends/backend_utils.py#L97

### PR DESCRIPTION
**Pull request by Michaelvll**
Fixes #3187
We reduce the CPU requirement for each spot controller process to allow 2x more spot jobs to be submitted to a spot controller.

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] Launch 32 spot jobs and manually kill all the instance to check if there is any OOM issue on the controller.
  - [x] `sky spot cancel -a`
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

**Review by concretevitamin — COMMENTED** [review]
Thanks, quick look.

File: sky/backends/backend_utils.py [hunk 1]
```diff
@@ -2468,8 +2467,8 @@ def get_task_demands_dict(task: 'task_lib.Task') -> Dict[str, float]:
     # For sky serve controller task, we set the CPU resource to a smaller
     # value to support a larger number of services.
     resources_dict = {
-        'CPU': (serve_lib.SERVICES_TASK_CPU_DEMAND
-                if task.service_name is not None else DEFAULT_TASK_CPU_DEMAND)
+        'CPU': (constants.CONTROLLER_PROCESS_CPU_DEMAND
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/backends/backend_utils.py:2472]: Good catch! Updated. Thanks!

File: sky/skylet/constants.py [hunk 2]
```diff
@@ -107,3 +107,11 @@
 # controller and sky serve controller.
 # TODO(tian): Refactor to controller_utils. Current blocker: circular import.
 CONTROLLER_IDLE_MINUTES_TO_AUTOSTOP = 10
+
+# Due to the CPU/memory usage of the controller process launched with sky job (
+# use ray job under the hood), we need to reserve some CPU/memory for each spot/
+# serve controller process.
+# Spot: A default controller with 8 vCPU can manage up to 32 spot jobs.
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/skylet/constants.py]: Added. Thanks!

File: sky/spot/constants.py [hunk 3]
```diff
@@ -9,8 +9,10 @@
 # Use default CPU instance type for spot controller, i.e.
 # m6i.2xlarge (8vCPUs, 32 GB) for AWS, Standard_D8s_v4 (8vCPUs, 32 GB)
 # for Azure, and n1-standard-8 (8 vCPUs, 32 GB) for GCP.
+# Based on profiling, the memory should be at least 3x than the CPU cores (we
+# allow 4x CPUs spot controller processes on the spot controller) to avoid OOM.
 # We use 50 GB disk size to reduce the cost.
-CONTROLLER_RESOURCES = {'disk_size': 50}
+CONTROLLER_RESOURCES = {'memory': '3x', 'disk_size': 50}
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/spot/constants.py:15]: Yes, it is 3x+. This `3x` is not exposed to the user, but indeed we should make it more clear in the future.

File: sky/spot/constants.py [hunk 4]
```diff
@@ -9,8 +9,10 @@
 # Use default CPU instance type for spot controller, i.e.
 # m6i.2xlarge (8vCPUs, 32 GB) for AWS, Standard_D8s_v4 (8vCPUs, 32 GB)
 # for Azure, and n1-standard-8 (8 vCPUs, 32 GB) for GCP.
+# Based on profiling, the memory should be at least 3x than the CPU cores (we
+# allow 4x CPUs spot controller processes on the spot controller) to avoid OOM.
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/spot/constants.py]: Rephrased it. PTAL : )

File: sky/spot/constants.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (concretevitamin) [reply sky/spot/constants.py:15]: Got it. Not sure I understand why the comment `(each spot controller process use 0.25 CPU cores)` connects to the explanation of memory?

**Review by concretevitamin — APPROVED** [review]
LGTM, thanks @Michaelvll.

File: sky/spot/constants.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/spot/constants.py:15]: Ahh, I further rephrase it to: `each CPU core can have 4 spot controller processes as we set the CPU requirement to 4, and 3 GB is barely enough for 4 spot processes`
Does that make it more clear?

[the remark this answers is not in the document]
Comment (concretevitamin) [reply sky/spot/constants.py:16]: It's clear now! Can we be consistent to say vCPU vs. CPU core (which is typically 2 vCPUs) everywhere?

**Review by concretevitamin — APPROVED** [review]
